design/aws-ecs-autoscaling-queue-worker.md

```diff
@@ -0,0 +1,225 @@
+
+The `Ec2QueueWorkerServiceProps` interface will contain properties to construct the Ec2TaskDefinition, SQSQueue and Ec2Service:
+
+```ts
+/**
+ * Properties to define an ECS service
+ */
+export interface Ec2QueueWorkerServiceProps {
+  /**
+   * The minimum number of CPU units to reserve for the container.
+   *
+   * @default none
+   */
+  readonly cpu?: number;
+
+  /**
+   * The hard limit (in MiB) of memory to present to the container.
+   *
+   * If your container attempts to exceed the allocated memory, the container
+   * is terminated.
+   *
+   * At least one of memoryLimitMiB and memoryReservationMiB is required for non-Fargate services.
+   */
+  readonly memoryLimitMiB?: number;
+
+  /**
+   * The soft limit (in MiB) of memory to reserve for the container.
+   *
+   * When system memory is under contention, Docker attempts to keep the
+   * container memory within the limit. If the container requires more memory,
+   * it can consume up to the value specified by the Memory property or all of
+   * the available memory on the container instance—whichever comes first.
+   *
+   * At least one of memoryLimitMiB and memoryReservationMiB is required for non-Fargate services.
+   */
+  readonly memoryReservationMiB?: number;
+}
+```
+
+An example use case:
+```ts
+// Create the vpc and cluster used by the Queue Worker task
+const vpc = new ec2.VpcNetwork(stack, 'Vpc', { maxAZs: 1 });
+const cluster = new ecs.Cluster(stack, 'EcsCluster', { vpc });
+cluster.addCapacity('DefaultAutoScalingGroup', {
+  instanceType: new ec2.InstanceType('t2.micro')
+});
+
+// Create the Queue Worker  task
+new Ec2QueueWorkerService(stack, 'EcsQueueWorkerService', {
+  cluster,
+  image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
+  desiredTaskCount: 2,
+  maxScalingCapacity: 5,
+  memoryReservationMiB: 512,
+  cpu: 256,
+  queueName: 'EcsWorkerQueue'
+});
+```
+
+# Part 3: Create `FargateQueueWorkerServiceProps` interface  and `FargateQueueWorkerService` construct  
+
+The `FargateQueueWorkerServiceProps` interface will contain properties to construct the FargateTaskDefinition, SQSQueue and FargateService:
+
+```ts
+/**
+ * Properties to define an Fargate service
+ */
+export interface FargateQueueWorkerServiceProps {
+    /**
+   * The number of cpu units used by the task.
+   * Valid values, which determines your range of valid values for the memory parameter:
+   * 256 (.25 vCPU) - Available memory values: 0.5GB, 1GB, 2GB
+   * 512 (.5 vCPU) - Available memory values: 1GB, 2GB, 3GB, 4GB
+   * 1024 (1 vCPU) - Available memory values: 2GB, 3GB, 4GB, 5GB, 6GB, 7GB, 8GB
+   * 2048 (2 vCPU) - Available memory values: Between 4GB and 16GB in 1GB increments
+   * 4096 (4 vCPU) - Available memory values: Between 8GB and 30GB in 1GB increments
+   *
+   * This default is set in the underlying FargateTaskDefinition construct.
+   *
+   * @default 256
+   */
+  readonly cpu?: string;
```

Comment:
cpu appears to be in both the Ec2 and Fargate queue workers -- could this be abstracted to the base construct?

Reply:
cpu is a string in Fargate (TaskDefProps accepts it as a string) and a number in Ec2 (ContainerDefProps accepts it as a number)